### PR TITLE
Spec + Conditionals: relative and absolute keynames

### PR DESCRIPTION
**Pull request by tom-wa**
This PR breaks the current behaviour for conditionals and mathcheck.

### CODE REVIEW

**Comment by markus2330** [comment]
Thanks for the PR! Updates in the README.md are however missing!

File: src/plugins/conditionals/conditionals.c [hunk 1]
```diff
@@ -187,7 +187,7 @@ static int compareStrings (const char * s1, const char * s2, const Key * suffixL
 	return retval;
 }
 
-static CondResult evalCondition (const char * leftSide, Comparator cmpOp, const char * rightSide, const char * condition,
+static CondResult evalCondition (Key * curKey, const char * leftSide, Comparator cmpOp, const char * rightSide, const char * condition,
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: why not const?

File: src/plugins/conditionals/conditionals.c [hunk 2]
```diff
@@ -218,14 +218,26 @@ static CondResult evalCondition (const char * leftSide, Comparator cmpOp, const
...
 			if (elektraRealloc ((void **)&lookupName, len) < 0)
 			{
 				ELEKTRA_SET_ERROR (87, parentKey, "Out of memory");
 				result = ERROR;
 				goto Cleanup;
 			}
-			snprintf (lookupName, len, "%s/%s", keyName (parentKey), rightSide);
+			if (rightSide[0] == '@')
+				snprintf (lookupName, len, "%s/%s", keyName (parentKey), rightSide + 1);
+			else if (rightSide[0] == '.')
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: inconsistent usage, why not using `!strncmp (rightSide, ".", 1)` again?

Some lines of comments would be very useful.

File: src/plugins/conditionals/conditionals.c [hunk 3]
```diff
@@ -380,7 +402,7 @@ static char * condition2cmpOp (const char * condition, Comparator * cmpOp)
 	return opStr;
 }
 
-static CondResult parseSingleCondition (const char * condition, const Key * suffixList, KeySet * ks, Key * parentKey)
+static CondResult parseSingleCondition (Key * key, const char * condition, const Key * suffixList, KeySet * ks, Key * parentKey)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: again, should be const.

File: src/plugins/conditionals/conditionals.c [hunk 4]
```diff
@@ -475,34 +497,65 @@ static const char * isAssign (char * expr, Key * parentKey, KeySet * ks)
 		--lastPtr;
 	if (*firstPtr != '\'' || *lastPtr != '\'')
 	{
-		if (lastPtr <= firstPtr) return NULL;
+		if (lastPtr <= firstPtr)
+		{
+			ELEKTRA_SET_ERRORF (134, parentKey,
+					    "Invalid syntax: \"%s\". Check kdb info conditionals for additional information\n", expr);
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: no newline at end of error text

File: src/plugins/conditionals/conditionals.c [hunk 5]
```diff
@@ -475,34 +497,65 @@ static const char * isAssign (char * expr, Key * parentKey, KeySet * ks)
...
+			return NULL;
+		}
 		char * nextMark = strchr (firstPtr + 1, '\'');
-		if (nextMark != lastPtr) return NULL;
+		if (nextMark != lastPtr)
+		{
+			ELEKTRA_SET_ERRORF (134, parentKey,
+					    "Invalid syntax: \"%s\". Check kdb info conditionals for additional information\n", expr);
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: Should these errors really have the same reason? Can you give the user some more specific information?

File: src/plugins/conditionals/testmod_conditionals.c [hunk 6]
```diff
@@ -261,7 +261,7 @@ static void test_doesntExistFail ()
 {
 	Key * parentKey = keyNew ("user/tests/conditionals", KEY_VALUE, "", KEY_END);
 	KeySet * ks = ksNew (5, keyNew ("user/tests/conditionals/totest", KEY_VALUE, "5", KEY_META, "check/condition",
-					"(totest<bla/val1) ? (! bla/val1)", KEY_END),
+					"(../totest<../bla/val1) ? (! ../bla/val1)", KEY_END),
```
Comment (markus2330) [inline src/plugins/conditionals/testmod_conditionals.c]: there is no testcase with `../../`

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
test from jenkins

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
@KurtMi Please review.

**Comment by markus2330** [comment]
@tom-wa please update as discussed

**Comment by markus2330** [comment]
@KurtMi @tom-wa Is this ready to merge?

**Comment by tom-wa** [comment]
did some tests, seems to work for me. a second opinion would be nice though

**Comment by KurtMi** [comment]
I tested also, looks fine

**Comment by markus2330** [comment]
@tom-wa thank you! Note that it was merged *after* the release.

**Pull request merged**